glide-core/redis-rs/redis/src/cluster_async/mod.rs

```diff
@@ -2713,36 +2739,81 @@ where
     }
 
     fn poll_recover(&mut self, cx: &mut task::Context<'_>) -> Poll<Result<(), RedisError>> {
-        trace!("entered poll_recovere");
+        trace!("entered poll_recover");
+
         let recover_future = match &mut self.state {
             ConnectionState::PollComplete => return Poll::Ready(Ok(())),
             ConnectionState::Recover(future) => future,
         };
+
         match recover_future {
-            RecoverFuture::RecoverSlots(ref mut future) => match ready!(future.as_mut().poll(cx)) {
-                Ok(_) => {
-                    trace!("Recovered!");
-                    self.state = ConnectionState::PollComplete;
-                    Poll::Ready(Ok(()))
-                }
-                Err(err) => {
-                    trace!("Recover slots failed!");
-                    let next_state = if err.kind() == ErrorKind::AllConnectionsUnavailable {
-                        ConnectionState::Recover(RecoverFuture::Reconnect(Box::pin(
-                            ClusterConnInner::reconnect_to_initial_nodes(self.inner.clone()),
-                        )))
-                    } else {
-                        ConnectionState::Recover(RecoverFuture::RecoverSlots(Box::pin(
-                            Self::refresh_slots_and_subscriptions_with_retries(
+            RecoverFuture::RefreshingSlots(result_receiver) => {
+                // Check if there's a result ready on the channel
+                match result_receiver.try_recv() {
```

Comment:
Instead of using a channel to get the result, we can store the refresh slots spawned task handle and then use it here to call handle.now_or_never(), which immediately returns with None when the task is still pending, and the result is being consumed when it's ready. It will reduce some of the error handling added due to additional channel, as with this now or never API we can communicate directly with the task itself (e.g you don't need to check if the channel is dropped and call refresh slots again).